sno/dataset1.py

```diff
@@ -218,68 +218,70 @@ def tupleizer(pk, blob):
+                    self.decode_pk(blob.name) if c == -1 else bin_feature[c]
                     for c in ftuple_order
                 ]
             )
 
         return tupleizer
 
-    def _features(self, feature_builder, fast):
-        top_tree = self.tree / ".sno-table"
+    def _iter_feature_blobs(self, fast=False):
+        """
+        Iterates over all the features in self.tree that match the expected
+        pattern for a feature, and yields the following for each:
+        >>> feature_builder(path_name, path_data)
+        """
+        sno_table_tree = self.tree / ".sno-table"
 
         # .sno-table/
         #   [hex(pk-hash):2]/
         #     [hex(pk-hash):2]/
         #       [base64(pk-value)]=[msgpack({attribute-id: attribute-value, ...})]
         URLSAFE_B64 = r"A-Za-z0-9_\-"
 
-        RE_L = re.compile(r"([0-9a-f]{2})?$")
-        RE_F = re.compile(
+        RE_DIR = re.compile(r"([0-9a-f]{2})?$")
+        RE_LEAF = re.compile(
             fr"(?:[{URLSAFE_B64}]{{4}})*(?:[{URLSAFE_B64}]{{2}}==|[{URLSAFE_B64}]{{3}}=)?$"
         )
 
-        for l1e in top_tree:
-            if l1e.type != pygit2.GIT_OBJ_TREE or not RE_L.match(l1e.name):
+        for dir1 in sno_table_tree:
+            if hasattr(dir1, "data") or not RE_DIR.match(dir1.name):
                 continue
 
-            for l2e in l1e:
-                if l2e.type != pygit2.GIT_OBJ_TREE or not RE_L.match(l2e.name):
+            for dir2 in dir1:
+                if hasattr(dir2, "data") or not RE_DIR.match(dir2.name):
                     continue
 
-                for fbe in l2e:
+                for leaf in dir2:
                     if not fast:
-                        if not RE_F.match(fbe.name):
+                        if not RE_LEAF.match(leaf.name):
                             continue
-                        elif fbe.type != pygit2.GIT_OBJ_BLOB:
+                        elif not hasattr(leaf, "data"):
+                            path = f".sno-table/{dir1.name}/{dir2.name}/{leaf.name}"
                             self.L.warn(
-                                "features: Unexpected TreeEntry type=%s in feature tree '%s/%s/%s'",
-                                fbe.type_str,
-                                l1e.name,
-                                l2e.name,
-                                fbe.name,
+                                f"features: No data found at path {path}, type={type(leaf)}"
                             )
                             continue
 
-                    yield feature_builder(fbe.name, fbe)
+                    yield leaf
 
     def features(self, *, ogr_geoms=False, **kwargs):
-        """ Feature iterator yielding (feature-key, feature-dict) pairs """
-        return self._features(
-            lambda pk, blob: (
-                pk,
-                self.repo_feature_to_dict(pk, blob, ogr_geoms=ogr_geoms),
-            ),
-            fast=False,
+        """ Feature iterator yielding (pk, feature-dict) pairs """
```

Comment:
pk here could be a tuple itself?

Reply:
A: no, looks like it's an encoded PK value[s]

Reply:
It looks like I made two changes here:

1. Change "feature-key" to "pk" in the comment, in line with the variable name in the code.
But, this was actually misleading since the "pk" variable did not store a pk value, but the blob name (ie, an encoded pk value).
2. Refactored the code so that the variable name "pk" disappeared. (This is dataset v1 so not valueS.)

Will fix the docstring.